File: JLQrCodePractice/JLQrCodePractice/JLQrCodePractice/JLScanViewController.m
//
//  JLScanViewController.m
//  JLQrCodePractice
//
//  Created by JackLin on 2017/3/24.
//  Copyright © 2017年 JackLin. All rights reserved.
//

#import "JLScanViewController.h"
#import <AVFoundation/AVFoundation.h>

@interface JLScanViewController ()<AVCaptureMetadataOutputObjectsDelegate, UINavigationControllerDelegate, UIImagePickerControllerDelegate>{
    
    AVCaptureDevice *_device;
    AVCaptureDeviceInput *_deviceInput;
    AVCaptureMetadataOutput *_metadataOutput;
    AVCaptureSession *_session;
    AVCaptureVideoPreviewLayer *_previewLayer;
    
}

@property (nonatomic, strong) UIButton *backButton;
@property (nonatomic, strong) UILabel *hintLabel;
@property (nonatomic, strong) UIImageView *scanImageView;
@property (nonatomic, strong) UILabel *scanHintLabel;
@property (nonatomic, strong) UILabel *resultLabel;
@end

@implementation JLScanViewController
#pragma mark - Life Cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"扫一扫";
    self.view.backgroundColor = [UIColor whiteColor];
    [self p_configueNavigationBar];
    [self p_isAvailableCamera];
}

- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
    if (_session) {
        [_session startRunning];
    }
}

- (void)viewWillDisappear:(BOOL)animated{
    [super viewWillDisappear:animated];
    if (_session) {
        [_session stopRunning];
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Getters
- (UIButton *)backButton{
    if (!_backButton) {
        _backButton = [UIButton buttonWithType:UIButtonTypeCustom];
        _backButton.frame = CGRectMake(0, 0, 62, 62);
        [_backButton setImage:[UIImage imageNamed:@"btn_return"] forState:UIControlStateNormal];
        [_backButton setContentHorizontalAlignment:UIControlContentHorizontalAlignmentLeft];
        [_backButton setImageEdgeInsets:UIEdgeInsetsMake(0, -12, 0, 0)];
        [_backButton setBackgroundColor:[UIColor clearColor]];
        [_backButton addTarget:self
                        action:@selector(onClickBackButton:)
              forControlEvents:UIControlEventTouchUpInside];
    }
    return _backButton;
}

- (UILabel *)hintLabel{
    if (!_hintLabel) {
        _hintLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, 200, 30)];
        _hintLabel.center = self.view.center;
        _hintLabel.font = [UIFont systemFontOfSize:16];
        _hintLabel.textColor = [UIColor orangeColor];
        _hintLabel.backgroundColor = [UIColor clearColor];
        _hintLabel.textAlignment = NSTextAlignmentCenter;
    }
    return _hintLabel;
}

- (UIImageView *)scanImageView{
    if (!_scanImageView) {
        _scanImageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, CGRectGetWidth(self.view.frame), CGRectGetHeight(self.view.frame))];
        _scanImageView.image = [UIImage imageNamed:@"img_qrCode"];
    }
    return _scanImageView;
}

- (UILabel *)scanHintLabel{
    if (!_scanHintLabel) {
        _scanHintLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, CGRectGetHeight(self.view.frame)/2 + 120, CGRectGetWidth(self.view.frame), 40)];
        _scanHintLabel.font = [UIFont systemFontOfSize:16];
        _scanHintLabel.textColor = [UIColor orangeColor];
        _scanHintLabel.backgroundColor = [UIColor clearColor];
        _scanHintLabel.textAlignment = NSTextAlignmentCenter;
        _scanHintLabel.text = @"将二维码/条码放在框内，即可自动扫描";
    }
    return _scanHintLabel;
}

- (UILabel *)resultLabel{
    if (!_resultLabel) {
        _resultLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, CGRectGetMaxY(self.scanHintLabel.frame) + 10,  CGRectGetWidth(self.view.frame), 80)];
        _resultLabel.backgroundColor = [UIColor clearColor];
        _resultLabel.font = [UIFont systemFontOfSize:16];
        _resultLabel.textColor = [UIColor grayColor];
        _resultLabel.textAlignment = NSTextAlignmentCenter;
        _resultLabel.numberOfLines = 0;
    }
    return _resultLabel;
}
#pragma mark - Action
- (void)onClickBackButton:(UIButton *)sender{
    [self.navigationController popViewControllerAnimated:YES];
}

- (void)onClickPhotoBarItem:(UIBarButtonItem *)sender{
    UIImagePickerController *imagePickerVirewController = [[UIImagePickerController alloc] init];
    imagePickerVirewController.allowsEditing = YES;
    imagePickerVirewController.delegate = self;
    imagePickerVirewController.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
    [self presentViewController:imagePickerVirewController animated:YES completion:^{
        if (_session) {
            [_session stopRunning];
        }
    }];
}

#pragma mark - Private
- (void)p_configueNavigationBar{
    self.navigationController.navigationBar.translucent = YES;
    UIBarButtonItem *leftBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:self.backButton];
    self.navigationItem.leftBarButtonItem = leftBarButtonItem;
    UIBarButtonItem*rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"相册"
                                                                          style:UIBarButtonItemStylePlain
                                                                         target:self
                                                                         action:@selector(onClickPhotoBarItem:)];
    self.navigationItem.rightBarButtonItem = rightBarButtonItem;
}

- (void)p_isAvailableCamera{
    BOOL isAvailable = [UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera];
    if (isAvailable) {
        [self p_getCameraAuthorizationStatus];
    }else{
        //摄像头不支持
        [self p_showCameraNotAvailableView];
    }
}

- (void)p_getCameraAuthorizationStatus{
    
    AVAuthorizationStatus authStatus = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeVideo];
    if (authStatus == AVAuthorizationStatusNotDetermined) {
        __weak typeof(self) weakSelf = self;
        [AVCaptureDevice requestAccessForMediaType:AVMediaTypeVideo completionHandler:^(BOOL granted) {
            if (granted) {
                //授权成功
                dispatch_async(dispatch_get_main_queue(), ^{
                    [weakSelf p_startScanningQrCode];
                    [weakSelf p_setupCustomView];
                });
                
            }else{
                //授权失败
                dispatch_async(dispatch_get_main_queue(), ^{
                     [self p_showNotDeterminedView];
                });
               
            }
        }];
    }else if (authStatus == AVAuthorizationStatusAuthorized){
        //已授权
        [self p_startScanningQrCode];
        [self p_setupCustomView];
    }else{
        //受限，未授权
        [self p_showNotDeterminedView];
    }
}

- (void)p_showCameraNotAvailableView{
    [self.view addSubview:self.hintLabel];
    self.hintLabel.text = @"摄像头不可用";
}

- (void)p_showNotDeterminedView{
    [self.view addSubview:self.hintLabel];
    self.hintLabel.text = @"未授权摄像头";
}

- (void)p_startScanningQrCode{
    _device = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
    
    if (!_session){
        _session = [[AVCaptureSession alloc] init];
    }
    
    NSError *error;
    _deviceInput = [AVCaptureDeviceInput deviceInputWithDevice:_device error:&error];
    if ([_session canAddInput:_deviceInput]) {
        [_session addInput:_deviceInput];
    }else{
        if (error) {
            NSLog(@"%@", error.description);
        }
    }
    
    if ([[[UIDevice currentDevice] systemVersion] floatValue] >= 7.0) {
        //7.0以上系统支持二维码扫描
        
        //元数据输出对象
        _metadataOutput = [[AVCaptureMetadataOutput alloc]init];
        //设置有效扫描区域
        [_metadataOutput setRectOfInterest:CGRectMake(0.22,0.16,0.43,0.68)];
        [_metadataOutput setMetadataObjectsDelegate:self queue:dispatch_get_main_queue()];
        //设置质量等级
        [_session setSessionPreset:AVCaptureSessionPresetHigh];
        if ([_session canAddOutput:_metadataOutput]) {
            [_session addOutput:_metadataOutput];
        }
        
        //设置条码类型
        _metadataOutput.metadataObjectTypes = @[AVMetadataObjectTypeQRCode];
        //添加扫描页面
        _previewLayer = [AVCaptureVideoPreviewLayer layerWithSession:_session];
        _previewLayer.videoGravity = AVLayerVideoGravityResizeAspectFill;
        _previewLayer.frame = self.view.layer.bounds;
        [self.view.layer insertSublayer:_previewLayer atIndex:0];
        [_session startRunning];
        
    }else{
        //考虑使用第三方库
    }
    
}

- (void)p_setupCustomView{
    [self.view addSubview:self.scanImageView];
    [self.view addSubview:self.scanHintLabel];
    [self.view addSubview:self.resultLabel];
}

- (NSString *)p_decodeImage:(UIImage *)image{
    NSData *imageData = UIImagePNGRepresentation(image);
    CIImage *ciimage = [CIImage imageWithData:imageData];
    if (!ciimage) {
        return nil;
    }
    CIDetector *qrDetector = [CIDetector detectorOfType:CIDetectorTypeQRCode
                                                context:[CIContext contextWithOptions:@{kCIContextUseSoftwareRenderer: @(YES)}]
                                                options:@{CIDetectorAccuracy : CIDetectorAccuracyHigh}];
    NSArray *resultArray = [qrDetector featuresInImage:ciimage];
    if (!resultArray.count) {
        return nil;
    }
    
    CIFeature *feature = resultArray.firstObject;
    CIQRCodeFeature *qrFeature = (CIQRCodeFeature *)feature;
    NSString *result = qrFeature.messageString;
    return result;
}

#pragma mark - AVCaptureMetadataOutputObjectsDelegate
- (void)captureOutput:(AVCaptureOutput *)captureOutput didOutputMetadataObjects:(NSArray *)metadataObjects fromConnection:(AVCaptureConnection *)connection{
    //扫描有结果，就停止扫描
    if ([metadataObjects count] > 0) {
        [_session stopRunning];
        
        AVMetadataMachineReadableCodeObject *metaDataObject = metadataObjects.firstObject;
        NSString *result = metaDataObject.stringValue;
        NSLog(@"result = %@", result);
        
        self.resultLabel.text = [NSString stringWithFormat:@"扫描结果：%@", result];
    }
}

#pragma mark - UIImagePickerControllerDelegate
- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary *)info{
    
    UIImage *selectedImage = info[@"UIImagePickerControllerEditedImage"];
    NSString *result = [self p_decodeImage:selectedImage];
    __weak typeof(self) weakSelf = self;
    [self dismissViewControllerAnimated:YES completion:^{
        if (result) {
            weakSelf.resultLabel.text = [NSString stringWithFormat:@"扫描结果：%@", result];
        }
    }];
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker{
    if (_session) {
        [_session startRunning];
    }
}

@end

File: JLQrCodePractice/JLQrCodePractice/JLScanViewController.swift
//
//  JLScanViewController.swift
//  JLQrCodePractice
//
//  Created by Sunxiaolin on 17/3/25.
//  Copyright © 2017年 JackLin. All rights reserved.
//

import UIKit
import AVFoundation

class JLScanViewController: UIViewController {
    
    private var captureDevice: AVCaptureDevice?
    private var deviceInput: AVCaptureDeviceInput?
    private var metadataOutput: AVCaptureMetadataOutput?
    fileprivate var captureSession: AVCaptureSession!
    private var previewLayer: AVCaptureVideoPreviewLayer?

    private lazy var backButton:UIButton! = {
        var button = UIButton(type: UIButtonType.custom)
        button.frame = CGRect(x: 0, y: 0, width: 62, height: 62)
        button.backgroundColor = UIColor.clear
        button.setTitle("扫一扫", for: UIControlState.normal)
        button.setTitleColor(UIColor.orange, for: UIControlState.normal)
        button.backgroundColor = UIColor.lightGray
        button.addTarget(self, action: #selector(onClickLeftBarItem(sender:)), for: UIControlEvents.touchUpInside)
        button.layer.borderWidth = 2.0
        button.layer.cornerRadius = 3.0
        button.clipsToBounds = true
        
        return button
    }()
    
    private lazy var hintLabel:UILabel! = {
        var label = UILabel(frame: CGRect(x: 0, y: 0, width: 200, height: 30))
        label.center = self.view.center
        label.font = UIFont.systemFont(ofSize: 16)
        label.textColor = UIColor.orange
        label.backgroundColor = UIColor.clear
        label.textAlignment = NSTextAlignment.center
        return label
    }()
    
    private lazy var scanImageView:UIImageView! = {
        var imageView = UIImageView(image: UIImage(named: "img_qrCode"))
        imageView.frame = CGRect(x: 0, y: 0, width: self.view.frame.width, height: self.view.frame.height)
        return imageView
    }()
    
    private lazy var scanHintLabel:UILabel! = {
        var label = UILabel(frame: CGRect(x: 0, y: self.view.frame.height/2 + 120, width: self.view.frame.width, height: 40))
        label.center = self.view.center
        label.font = UIFont.systemFont(ofSize: 16)
        label.textColor = UIColor.orange
        label.backgroundColor = UIColor.clear
        label.textAlignment = NSTextAlignment.center
        label.text = "将二维码/条码放在框内，即可自动扫描"
        return label
    }()
    
    fileprivate lazy var resultLabel:UILabel! = {
        var label = UILabel(frame: CGRect(x: 0, y: self.scanHintLabel.frame.maxY + 10, width: self.view.frame.width, height: 80))
        label.font = UIFont.systemFont(ofSize: 16)
        label.textColor = UIColor.orange
        label.backgroundColor = UIColor.clear
        return label
    }()
    
    
    //MARK: Life Cycle
    override func viewDidLoad() {
        super.viewDidLoad()

        self.title = "扫一扫"
        self.view.backgroundColor = UIColor.white
        configueNavigationBar()
        isAvailableCamera()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated);
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    //MARK: Private
    private func configueNavigationBar(){
        
        let leftBarButtonItem = UIBarButtonItem(image: UIImage(named: "btn_return"), style: UIBarButtonItemStyle.plain, target: self, action: #selector(onClickLeftBarItem(sender:)))
        self.navigationItem.leftBarButtonItem = leftBarButtonItem
        
        let rightBarButtonItem = UIBarButtonItem(title: "相册", style: UIBarButtonItemStyle.plain, target: self, action: #selector(onClickRightBarItem(sender:)))
        self.navigationItem.rightBarButtonItem = rightBarButtonItem
    }
    
    func isAvailableCamera(){
        
        let isAvailable = UIImagePickerController.isSourceTypeAvailable(UIImagePickerControllerSourceType.camera)
        if isAvailable{
            getCameraAuthorizationStatus()
        }else{
            showCameraNotAvailableView()
        }
        
    }
    
    func getCameraAuthorizationStatus(){
        let authStatus = AVCaptureDevice.authorizationStatus(forMediaType: AVMediaTypeVideo)
        if authStatus == AVAuthorizationStatus.notDetermined{
            
            AVCaptureDevice.requestAccess(forMediaType: AVMediaTypeVideo, completionHandler: { (granted) -> Void in
                if granted{
                    DispatchQueue.main.async {
                        self.startScanningCode()
                        self.setupCustomView()
                    }
                    
                }else{
                    DispatchQueue.main.async {
                        self.showCameraNotDeterminedView()
                    }
                }
            })
            
        }else if authStatus == AVAuthorizationStatus.authorized{
            self.startScanningCode()
            self.setupCustomView()
            
        }else{
            showCameraNotDeterminedView()
        }
    }
    
    func startScanningCode(){
        captureDevice = AVCaptureDevice.defaultDevice(withMediaType: AVMediaTypeVideo)
        if (captureSession == nil){
            captureSession = AVCaptureSession()
        }
        do{
            deviceInput = try AVCaptureDeviceInput.init(device: captureDevice)
        }catch let error as NSError{
            print(error.description)
        }
        
        captureSession.addInput(deviceInput)
        
        let systemVer = Double(UIDevice.current.systemVersion)!

        if systemVer > 7.0 {
            
            metadataOutput = AVCaptureMetadataOutput()
            //set scan rectange
            metadataOutput?.rectOfInterest = CGRect(x: 0.22, y: 0.16, width: 0.43, height: 0.68)
            metadataOutput?.setMetadataObjectsDelegate(self, queue: DispatchQueue.main)
            captureSession?.canSetSessionPreset(AVCaptureSessionPresetHigh)
            if ((captureSession?.canAddOutput(metadataOutput)) != nil){
                captureSession?.addOutput(metadataOutput)
            }
            
            //set code type
            if #available(iOS 9.0, *) {
                metadataOutput?.metadataObjectTypes = [AVMediaTypeMetadataObject]
                
                previewLayer = AVCaptureVideoPreviewLayer(session: captureSession)
                previewLayer?.videoGravity = AVLayerVideoGravityResizeAspectFill
                previewLayer?.frame = self.view.bounds
                self.view.layer.insertSublayer(previewLayer!, at: 0)
                captureSession?.startRunning()
                
            } else {
                // Fallback on earlier versions
            }
            
            
        }else{
            //need third lib to supported
        }
    }
    
    func setupCustomView(){
        self.view.addSubview(self.scanImageView)
        self.view.addSubview(self.scanHintLabel)
        self.view.addSubview(self.resultLabel)
        
    }
    
    func showCameraNotAvailableView(){
        self.view.addSubview(self.hintLabel)
        self.hintLabel.text = "摄像头不可用"
    }
    
    func showCameraNotDeterminedView(){
        self.view.addSubview(self.hintLabel)
        self.hintLabel.text = "未授权摄像头"
    }
    
    func decodeImage(decodedImage:UIImage)->String{
        let data = UIImagePNGRepresentation(decodedImage)
        
        guard let ciimage = CIImage(data: data!) else{
            return ""
        }
        
        let qrDtector = CIDetector(ofType: CIDetectorTypeQRCode, context: CIContext(options: [kCIContextUseSoftwareRenderer: (true)]), options: [CIDetectorAccuracy: CIDetectorAccuracyHigh])
        
        let resultArray = qrDtector?.features(in: ciimage)
        if (resultArray?.count)! > 0
        {
            
            let feature = resultArray?.first
            let qrFeature = feature as! CIQRCodeFeature
            return qrFeature.messageString!
            
        }else{
            return ""
        }
        
    }
    
    
    
    //MARK: Acrion Method
    func onClickLeftBarItem(sender:UIBarButtonItem){
        self.navigationController?.popViewController(animated: true)
    }
    
    func onClickRightBarItem(sender:UIBarButtonItem){
        let imagePickerViewController = UIImagePickerController()
        imagePickerViewController.allowsEditing = true
        imagePickerViewController.delegate = self
        imagePickerViewController.sourceType = UIImagePickerControllerSourceType.photoLibrary
        self.present(imagePickerViewController, animated: true) { () -> Void in
            if self.captureSession != nil{
                self.captureSession?.stopRunning()
            }
        }
    }
    

}

extension JLScanViewController:AVCaptureMetadataOutputObjectsDelegate{
    
    func captureOutput(_ captureOutput: AVCaptureOutput!, didOutputMetadataObjects metadataObjects: [Any]!, from connection: AVCaptureConnection!){
        
        if metadataObjects.count > 0{
            self.captureSession?.stopRunning()
            
            let metadataObject = metadataObjects.first as! AVMetadataMachineReadableCodeObject
            
            guard let result = metadataObject.stringValue else{
                return
            }
            
            print(result)
            self.resultLabel.text = result
        }
    }
}

extension JLScanViewController:UINavigationControllerDelegate, UIImagePickerControllerDelegate{
    private func imagePickerController(picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : AnyObject]) {
        let selectedImage = info["UIImagePickerControllerEditedImage"] as! UIImage
        let result = self.decodeImage(decodedImage: selectedImage)
        
        print("result:\(result)")
        
        self.dismiss(animated: true) { () -> Void in
            if result != ""{
                self.resultLabel.text = result
            }
        }
        
    }
    
    func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
        if self.captureSession != nil{
            captureSession?.startRunning()
        }
    }
}
